### PR TITLE
refactor: Complete rewrite of the library

### LINKED ISSUE #38: Refactor validators
Currently, we have one huge function responsible for validation. We should split it into separate functions bundled together with specific validators. The aim is to solve two problems:

- [x] limited tree-shaking – make validation code tree-shakable
- [x] broken Single Responsibility Principle – when adding new validators we need to modify two places

This should not be a breaking change in the user-facing API.
Validators implementation details should be considered private.

### PR DESCRIPTION
**Pull request by mmiszy**
Resolves #38 

BREAKING CHANGES: There are **none** breaking changes in the user-facing APIs. However, internal structures used by the library have vastly changed and have been reorganized. As a result, schemas are no longer easily serialized to JSON.

### CODE REVIEW

File: src/validators/__schema.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+export const TYPEOFWEB_SCHEMA = Symbol('@typeofweb/schema');
+
+export const isSchema = (val: any): val is SomeSchema<any> => {
+  return (
+    typeof val === 'object' &&
+    val !== null &&
+    '__validator' in val &&
+    '__modifiers' in val &&
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    val[TYPEOFWEB_SCHEMA] === true
```
Comment (wisnie) [inline src/validators/__schema.ts:12]: Why do we have  to check ourselves? If `TYPEOFWEB_SCHEMA` is unique and only we can set this property. If we set this property to true, then isn't val always an object, `'__validator'` always on this object and `'__modifiers'` also always on this object?

File: src/validators/__stringifyHelpers.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import type { SomeSchema } from '../types';
+
+export const typeToPrint = (str: string) => '≫' + str + '≪';
```
Comment (wisnie) [inline src/validators/__stringifyHelpers.ts:3]: I'm afraid that this symbol's (`≫`) display depends on the installed font on the  machine that will run our library, so I think the display of this symbol may look ugly or not be displayed at all.

Comment (mmiszy) [reply src/validators/__stringifyHelpers.ts:3]: Any examples?

Comment (wisnie) [reply src/validators/__stringifyHelpers.ts:3]: image

File: src/validators/__validate.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import type { SomeSchema, TypeOf } from '..';
+import { ValidationError } from '../errors';
+import type { Either } from '../types';
+
+export const validate = <S extends SomeSchema<any>>(schema: S) => (value: unknown) => {
+  const result = __validate<S>(schema, value);
+  if (result._t === 'right') {
+    return result.value;
+  } else {
+    // throw result.value;
```
Comment (AdamSiekierski) [inline src/validators/__validate.ts:10]: Chyba wyrzucić można ten komentarz :)

File: src/validators/array.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+/* eslint-disable functional/no-this-expression */
```
Comment (AdamSiekierski) [inline src/validators/array.ts:1]: Może po prostu wyłączyć tę zasadę, bo w całkiem wielu plikach jest ona wyłączona

File: src/validators/boolean.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import { ValidationError } from '../errors';
+import { initialModifiers } from '../schema';
+import { typeToPrint } from '../stringify';
+import type { Schema } from '../types';
+
+export const boolean = () => {
+  return {
+    __modifiers: initialModifiers,
+    toString() {
+      return typeToPrint('boolean');
```
Comment (AdamSiekierski) [inline src/validators/boolean.ts:11]: Po co to? Ta funkcja zwraca po prostu `'boolean'` o ile się dobrze orientuję

File: src/validators/date.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import type { Schema } from '../types';
+import { isDate, isISODateString } from '../utils/dateUtils';
+
+import { __validate } from './__validate';
+
+export const date = () => {
+  return {
+    __modifiers: initialModifiers,
+    toString() {
+      return typeToPrint('Date');
```
Comment (AdamSiekierski) [inline src/validators/date.ts:14]: Tutaj tak samo 🤔

File: __tests__/exports.spec.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+/* eslint-disable import/dynamic-import-chunkname */
+import Fs from 'fs';
+import { resolve, extname } from 'path';
```
Comment (wisnie) [inline __tests__/exports.spec.ts:3]: Dlaczego  `Fs` w PascalCase?

Comment (mmiszy) [reply __tests__/exports.spec.ts:3]: Przyzwyczajenie z Hapi.

File: src/validators/boolean.ts (hunk 5 above)
Comment (mmiszy) [reply src/validators/boolean.ts:11]: Abstraction

File: src/validators/__validate.ts (hunk 3 above)
Comment (mmiszy) [reply src/validators/__validate.ts:10]: This will be used in #13

File: src/validators/array.ts (hunk 4 above)
Comment (mmiszy) [reply src/validators/array.ts:1]: Sounds good

File: src/modifiers/optional.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    S['__type'],
+    MergeModifiers<S['__modifiers'], { readonly optional: true }>,
+    S['__values']
+  >;
+};
+
+export const isOptionalSchema = (
+  s: SomeSchema<any>,
+): s is SomeSchema<any> & { readonly __modifiers: { readonly optional: true } } =>
+  !!s.__modifiers.optional;
```
Comment (wisnie) [inline src/modifiers/optional.ts:20]: It doesn't really matter, but I think we can use `Boolean` for clarity :)

File: src/stringify.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import type { SomeSchema } from './types';
+
+export const schemaToString = (schema: SomeSchema<any>): string => {
+  const modifiers = getModifiers(schema);
+  return unionToPrint([schema.toString(), ...modifiers]);
+};
+
+export const typeToPrint = (str: string) => str;
+export const objectToPrint = (str: string) => '{' + str + '}';
```
Comment (wisnie) [inline src/stringify.ts:9]: Why do we use `string concatenation` in this particular place and backticks in the other functions?

File: src/stringify.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  if (arr.length > 1) {
+    return `(${str})`;
+  }
+  return str;
+};
+
+const getModifiers = (v: SomeSchema<any>): readonly string[] => {
+  const modifiers = [v.__modifiers.optional && 'undefined', v.__modifiers.nullable && 'null'];
+  return modifiers.filter((m): m is string => Boolean(m));
+};
```
Comment (wisnie) [inline src/stringify.ts:24]: Won't we include `minLength`?

Comment (mmiszy) [reply src/stringify.ts:24]: For what?

Comment (wisnie) [reply src/stringify.ts:24]: Okay, you're right, but this function name confused me.

File: src/validators/array.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+function validateArray<
+  U extends readonly SomeSchema<unknown>[],
+  TypeOfResult extends readonly TypeOf<U[number]>[]
+>(this: Schema<TypeOfResult, typeof initialModifiers, U>, value: unknown) {
+  if (!Array.isArray(value)) {
+    return left(new ValidationError(this, value));
+  }
+
+  return sequenceA<readonly unknown[], readonly unknown[]>((val) => {
+    const isValid = this.__values.some((validator) => validator.__validate(val)._t === 'right');
```
Comment (mmiszy) [inline src/validators/array.ts:36]: Should be find

File: src/validators/oneOf.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
+import { ValidationError } from '../errors';
+import { initialModifiers, isSchema } from '../schema';
+import { schemaToString } from '../stringify';
+import type { Either, Primitives, Schema, SomeSchema, TypeOf } from '../types';
+import { left, right } from '../utils/either';
+
+// `U extends (Primitives)[]` and `[...U]` is a trick to force TypeScript to narrow the type correctly
+// thanks to schema, there's no need for "as const": oneOf(['a', 'b']) works as oneOf(['a', 'b'] as const)
```
Comment (mmiszy) [inline src/validators/oneOf.ts:8]:
```suggestion
// thanks to this, there's no need for "as const": oneOf(['a', 'b']) works as oneOf(['a', 'b'] as const)
```

**Pull request merged**